app/models/piece.rb

```diff
@@ -3,5 +3,145 @@ class Piece < ActiveRecord::Base
 	belongs_to :user
 	belongs_to :game
 
+	def executable_move?(x_target, y_target)
+		return false unless valid_move?(x_target, y_target) # => if the move isn't valid for its type 
+		return false unless on_board?(x_target, y_target) # => if the piece tries to move outside the board 
+		return false if obstacle_between?(x_target, y_target) # => if there is obstruction in between 
+		return false if no_move?(x_target, y_target) # => if the piece doesn't move at all 
+	end 
+
+	def no_move?(x_target, y_target) 
+		(x_position == x_target) && (y_position == y_target)
+	end 
+
+	def obstacle_between(x_target, y_target)
+		return false if diagonal_obstruction?(x_target, y_target).include?([x_target, y_target)
+		return false if rectilinear_collection?(x_target, y_target).include?([x_target, y_target])
+	end 
+
+	def on_board?(x_target, y_target)
+		false unless (0 <= x_target <= 7) && (0 <= y_target <= 7)
+	end 
+
```

Comment:
Add a new implementation of `diagonal_obstruction?` where it loops through all of the points from diagonal_path and looks for pieces at those points (e.g. `Piece.find_by_position(x, y, game_id)`)